import mongoose from "mongoose"
import { AnoDeclaracao, Declaracoes } from "../models"

async function createAnoToObjectIdMap() {
  const anoToObjectIdMap = new Map()

  try {
    const anosDeclaracao = await AnoDeclaracao.find({}, { ano: 1, _id: 1 })

    for (const doc of anosDeclaracao) {
      anoToObjectIdMap.set(doc.ano.toString(), doc._id)
    }

    console.log("‚úÖ Mapa de anos criado com sucesso!", anoToObjectIdMap)
  } catch (error) {
    console.error("Erro ao criar o mapa de anos:", error)
  }

  return anoToObjectIdMap
}

async function migrateAnoDeclaracao(anoToObjectIdMap) {
  try {
    const declaracoes = await Declaracoes.find({}, { _id: 1, anoDeclaracao: 1 })

    console.log(
      `üîÑ Iniciando migra√ß√£o para ${declaracoes.length} declara√ß√µes...`
    )

    for (const doc of declaracoes) {
      const { _id, anoDeclaracao } = doc

      if (
        typeof anoDeclaracao === "string" &&
        anoToObjectIdMap.has(anoDeclaracao)
      ) {
        const objectId = anoToObjectIdMap.get(anoDeclaracao)

        await Declaracoes.updateOne(
          { _id },
          { $set: { anoDeclaracao: objectId } }
        )

        console.log(
          `‚úÖ Declara√ß√£o ${_id} atualizada: ${anoDeclaracao} ‚ûù ${objectId}`
        )
      } else if (anoDeclaracao === undefined) {
        console.warn(`‚ö†Ô∏è Declara√ß√£o ${_id} tem anoDeclaracao undefined!`)
      } else {
        console.warn(
          `‚ö†Ô∏è Sem ObjectId para o ano: ${anoDeclaracao} (Declara√ß√£o ${_id})`
        )
      }
    }

    console.log("üéâ Migra√ß√£o conclu√≠da com sucesso!")
  } catch (error) {
    console.error("‚ùå Erro durante a migra√ß√£o:", error)
  }
}

async function main() {
  try {
    console.log("üîó Conectando ao banco de dados...")
    await mongoose.connect(
      "mongodb://root:asdf1234@mongo:27017/INBCM?authSource=admin"
    )

    console.log("‚úÖ Conex√£o estabelecida!")

    const anoToObjectIdMap = await createAnoToObjectIdMap()
    await migrateAnoDeclaracao(anoToObjectIdMap)
  } catch (error) {
    console.error("‚ùå Erro no processo principal:", error)
  } finally {
    await mongoose.disconnect()
    console.log("üîå Conex√£o encerrada.")
  }
}

main()
